### DISCUSSION TITLE
How to create plugins, as of Oct. 1, 2022?

### DISCUSSION THREAD

**Discussion by ehaynes99** [swc-project/swc#6013] [General]
I get that this is still an API in flux, but it's getting rather difficult to track the changes. I don't expect extensive documentation for something that's changing this frequently, but I haven't been able to make a "hello world" work that's still valid after a week. Is there a reasonable checklist of things to determine "how can I make a plugin that works?" If the answer is "go away while we settle it", that's fine too, but if so the docs should go back to calling it experimental.

Currently, if you follow the [docs](https://swc.rs/docs/plugin/ecmascript/getting-started) and create a plugin project via:
```
swc plugin new --target-type wasm32-wasi my-first-plugin
```
It creates a project with `Cargo.toml` including:
```
swc_core = { version = "0.20.*", features = ["plugin_transform"] }
```
attempting to `cargo build` on the resulting project hangs indefinitely on:
```
Building [======================>  ] 243/254: swc_ecma_ast
```
One of my CPUs stays pegged at 100%. After roughly 30s, it starts switching which core that is every 2s, but there's always one that's maxed, and it never completes. I've tried this on multiple machines, all Arch Linux with current toolchains.

---
Updating the `Cargo.toml` file to:
```
swc_core = { version = "0.28.*", features = ["plugin_transform"] }
```
There is an error:
```
error[E0432]: unresolved imports `swc_core::ast`, `swc_core::testing_transform`, `swc_core::visit`
 --> src/lib.rs:2:5
  |
2 |     ast::Program,
  |     ^^^ could not find `ast` in `swc_core`
3 |     plugin::{plugin_transform, proxies::TransformPluginProgramMetadata},
4 |     testing_transform::test,
  |     ^^^^^^^^^^^^^^^^^ could not find `testing_transform` in `swc_core`
5 |     visit::{as_folder, FoldWith, VisitMut},
  |     ^^^^^ could not find `visit` in `swc_core`
```

---
The example plugin in those docs hasn't compiled for a while now either:
```rust
fn visit_mut_bin_expr(&mut self, e: &mut BinExpr) {
    e.visit_mut_children_with(self);

    if e.op == op!("===") {
        e.left = Box::new(Ident::new("kdy1".into(), e.left.span()).into());
    }
}
```
Last time I could get something to compile at all, I \*think\* this is what would work now:
```rust
if e.op == op!("===") {
    e.left = Box::new(Expr::Ident(Ident::new("kdy1".into(), e.left.span())));
}
```

I'd rank myself a 9/10 for TypeScript, but a 2/10 for Rust, so I assume some of the toolchain issues are more obvious to Rustaceans, but is anyone interested in creating a repo with release tags for at least that "getting started" plugin that stays current?

**Comment by kwonoj** [comment]
The main problem in here is we are behind of updating swc_cli's template after we make some changes.

Most straightforward for now is referring latest plugins, like https://github.com/swc-project/plugins/blob/9054248ecde9114521cf634b06fd61f51355f535/packages/jest/src/lib.rs#L5-L9 which have correct imports for the changed namespaces, also need to use latest swc_core if possible.
